### PR TITLE
Core: Add support for `beforeAll` hook

### LINKED ISSUE #28268: Add beforeAll hook to Storybook
The `beforeAll` hook can be defined on a preview level, or implicitly via addons (in CSF we would say projectAnnotations). First run the addon's beforeAll, and then the preview beforeAll.

usage:

```ts
// .storybook/preview

const preview = {
  beforeAll: async () => {
    console.log('beforeAll')

    // triggered via HMR when preview.ts is changed
    return () => console.log('beforeAll cleanup')
  }
}

export default preview
```

for portable stories, the hook would be part of the project annotations:
```ts
// setupTest.ts
import { beforeAll } from 'vitest'
import { setProjectAnnotations } from '@storybook/react'
import * as previewAnnotations from '../.storybook/preview'

const storybookAnnotations = setProjectAnnotations(previewAnnotations)

beforeAll(storybookAnnotations.beforeAll)

// or
beforeAll(async () => {
  // other code here, plus beforeAll
  await storybookAnnotations.beforeAll()
})
```

### PR DESCRIPTION
**Pull request by ghengeveld**
Closes #28268
Depends on https://github.com/ComponentDriven/csf/pull/96 ✅

## What I did

This adds the concept of a `beforeAll` function, which runs when the Storybook preview is initialized or preview annotations are updated. In general, `beforeAll` runs once, before any loaders, decorators or stories. The API is similar to Vitest's [beforeAll](https://vitest.dev/api/#beforeall) (except that timeout is not supported):

```ts
type Awaitable<T> = T | Promise<T>;
type CleanupFn = () => Awaitable<void>;
type BeforeAll = () => Awaitable<void | CleanupFn>;
```

`beforeAll` may be synchronous or asynchronous, and may return a cleanup function which may also be synchronous or asynchronous.

`beforeAll` may be specified as an export in `preview.js` (by addons or in user's Storybook config) or in a file referenced by `previewAnnotations` in `main.js`.

```js
// .storybook/preview.js

export const beforeAll = () => {
  // ... do something

  return () => {
    // ... cleanup
  }
}
```

When multiple `beforeAll` hooks are specified, they will be executed sequentially (and awaited) in the following order: 

1. Addon hooks (in order of `addons` array in `.storybook/main.js`)
2. Annotation hooks (in order of `previewAnnotations` array in `.storybook/main.js`)
3. Preview hook (via `.storybook/preview.js`)

The cleanup function acts like an "afterAll", but is not guaranteed to run because this code typically runs in the browser. However, the Storybook test runner may invoke this function to clean up between test runs. Also, when modifying `preview.js` or a previewAnnotations file, cleanup will run before reinitializing `beforeAll`. Cleanup functions are executed sequentially, in reverse order of initialization:

1. Preview cleanup
2. Annotation cleanup (in reverse order)
3. Addon cleanup (in reverse order)

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

1. Create a sandbox
2. Create `.storybook/before-all.js` with the example beforeAll hook below.
3. Update `.storybook/preview.js` to add `beforeAll` (see below)
4. Update `.storybook/main.js` to add `previewAnnotations` (see below)
5. Run the sandbox and check your browser console
6. Verify you see both hooks run (`before-all.js` goes first)
7. Modify and save `preview.js`, then check your browser console again
8. Verify you see the cleanup happen in reverse order before the updated hooks run in their usual order.

Example `beforeAll` hook:

```js
// .storybook/preview.js or .storybook/before-all.js
export const beforeAll = () => {
  console.log('### beforeAll in preview/annotation');
  return () => {
    console.log('### afterAll in preview/annotation');
  };
};
```

Update the log lines so you can distinguish between files (preview.js, before-all.js).

```js
// .storybook/main.js
const config = {
  // ... other stuff

  previewAnnotations: ['.storybook/before-all.js']
};
export default config;
```

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-28255-sha-9d564a0d`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-28255-sha-9d564a0d sandbox` or in an existing project with `npx storybook@0.0.0-pr-28255-sha-9d564a0d upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-28255-sha-9d564a0d`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-28255-sha-9d564a0d) |
| **Triggered by** | @ghengeveld |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`beforeAll-hook`](https://github.com/storybookjs/storybook/tree/beforeAll-hook) |
| **Commit** | [`9d564a0d`](https://github.com/storybookjs/storybook/commit/9d564a0d0759575e954d0a74a83786ee0b281cb8) |
| **Datetime** | Mon Jun 24 13:56:48 UTC 2024 (`1719237408`) |
| **Workflow run** | [9646676768](https://github.com/storybookjs/storybook/actions/runs/9646676768) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=28255`_
</details>

### CODE REVIEW

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// Execute all the hooks in sequence, and return a function that will execute cleanups in reverse order
+export const composeBeforeHooks = (hooks: BeforeAllHook[]): BeforeAllHook => {
+  return async () => {
+    const cleanups: CleanupFn[] = [];
+    for (const hook of hooks) {
+      const cleanup = await hook();
+      if (cleanup) cleanups.push(cleanup);
+    }
+    return async () => {
+      for (const cleanup of cleanups.reverse()) {
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/store/csf/beforeAll.ts:16]: Maybe to be sure, [...cleanups.reverse()] as reverse is mutable.
Although technically, this should not be called twice.

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+type Awaitable<T> = T | Promise<T>;
+
+type CleanupFn = () => Awaitable<void>;
+
+export type BeforeAllHook = () => Awaitable<void | CleanupFn>;
+
+// Execute all the hooks in sequence, and return a function that will execute cleanups in reverse order
+export const composeBeforeHooks = (hooks: BeforeAllHook[]): BeforeAllHook => {
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/store/csf/beforeAll.ts:8]: `composeBeforeAllHooks`?

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+type Awaitable<T> = T | Promise<T>;
+
+type CleanupFn = () => Awaitable<void>;
+
+export type BeforeAllHook = () => Awaitable<void | CleanupFn>;
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/store/csf/beforeAll.ts:5]: Imported from CSF?

File: code/lib/preview-api/src/modules/store/csf/portable-stories.ts [hunk 4]
```diff
@@ -35,8 +35,10 @@ const DEFAULT_STORY_NAME = 'Unnamed Story';
 function extractAnnotation<TRenderer extends Renderer = Renderer>(
   annotation: NamedOrDefaultProjectAnnotations<TRenderer>
 ) {
+  if (!annotation) return {};
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/store/csf/portable-stories.ts:38]: why is this added?

File: code/lib/preview-api/src/modules/store/csf/portable-stories.ts [hunk 5]
```diff
@@ -165,7 +169,7 @@ export function composeStory<TRenderer extends Renderer = Renderer, TArgs extend
         cleanups.push(
           ...(await story.applyBeforeEach(context))
             .filter(Boolean)
-            .map((callback) => ({ storyName, callback }))
+            .map((callback: () => void) => ({ storyName, callback }))
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/store/csf/portable-stories.ts:172]: why is this added?

File: code/lib/types/src/modules/story.ts [hunk 6]
```diff
@@ -46,6 +46,8 @@ export type ProjectAnnotations<TRenderer extends Renderer> = CsfProjectAnnotatio
 
   /* @deprecated use renderToCanvas */
   renderToDOM?: RenderToCanvas<TRenderer>;
+
+  beforeAll?: () => MaybePromise<void | (() => MaybePromise<void>)>;
```
Comment (kasperpeulen) [inline code/lib/types/src/modules/story.ts:50]: move to CSF

File: code/renderers/svelte/src/portable-stories.ts [hunk 7]
```diff
@@ -10,6 +10,7 @@ import type {
   Store_CSFExports,
   StoriesWithPartialProps,
   ComposedStoryFn,
+  NamedOrDefaultProjectAnnotations,
```
Comment (kasperpeulen) [inline code/renderers/svelte/src/portable-stories.ts:13]: Why did we change to this type?

File: code/lib/preview-api/src/modules/preview-web/Preview.tsx [hunk 8]
```diff
@@ -160,9 +162,23 @@ export class Preview<TRenderer extends Renderer> {
...
-    this.projectAnnotationsBeforeInitialization = projectAnnotations;
+  async initializeWithProjectAnnotations(
+    projectAnnotations: ProjectAnnotations<TRenderer>,
+    storyStore?: StoryStore<TRenderer>
+  ) {
     try {
+      await this.beforeAllCleanup?.();
+      this.beforeAllCleanup = await projectAnnotations.beforeAll?.();
+    } catch (err) {
+      this.renderPreviewEntryError('Error in beforeAll hook:', err as Error);
```
Comment (kasperpeulen) [inline code/lib/preview-api/src/modules/preview-web/Preview.tsx:173]: How does such a preview entry error looks like?

**Review by kasperpeulen — COMMENTED** [review]
Looks good!

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts (hunk 1 above)
Comment (ghengeveld) [reply code/lib/preview-api/src/modules/store/csf/beforeAll.ts]: Sure. I've replaced `push` with `unshift` so we avoid reversing altogether.

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts (hunk 2 above)
Comment (ghengeveld) [reply code/lib/preview-api/src/modules/store/csf/beforeAll.ts:8]: Yeah I wasn't sure whether we'd want to reuse this for `beforeEach`. WDYT?

File: code/lib/preview-api/src/modules/store/csf/portable-stories.ts (hunk 5 above)
Comment (ghengeveld) [reply code/lib/preview-api/src/modules/store/csf/portable-stories.ts:172]: Doesn't seem necessary, I'll remove.

File: code/renderers/svelte/src/portable-stories.ts (hunk 7 above)
Comment (ghengeveld) [reply code/renderers/svelte/src/portable-stories.ts:13]: @yannbf ?

File: code/lib/preview-api/src/modules/preview-web/Preview.tsx (hunk 8 above)
Comment (ghengeveld) [reply code/lib/preview-api/src/modules/preview-web/Preview.tsx:173]: Whatever random thing you might be doing in your beforeAll hook. There's no telling what random shenanigans people may end up (ab)using this for.

File: code/lib/preview-api/src/modules/preview-web/Preview.tsx [hunk 9]
```diff
@@ -160,9 +162,23 @@ export class Preview<TRenderer extends Renderer> {
...
+      throw err;
+    }
+
+    // Don't reinitialize story store if it's already been set.
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply code/lib/preview-api/src/modules/preview-web/Preview.tsx:177]: This guard was previously at the callsite of `initializeWithProjectAnnotations` (`if (!this.storyStoreValue)` below), I merely moved it into the function.

The rationale is that this function now also invokes the `beforeAll` hook, and it should do that even if the story index has already been initialized (in which case we'll run cleanup, rerun beforeAll and skip initialization of the story index).

[the remark this answers is not in the document]
Comment (kasperpeulen) [reply code/lib/preview-api/src/modules/preview-web/Preview.tsx:177]: 👍

File: code/lib/preview-api/src/modules/preview-web/Preview.tsx (hunk 8 above)
Comment (kasperpeulen) [reply code/lib/preview-api/src/modules/preview-web/Preview.tsx:173]: Okay fair enough

File: code/lib/preview-api/src/modules/store/csf/beforeAll.ts (hunk 2 above)
Comment (kasperpeulen) [reply code/lib/preview-api/src/modules/store/csf/beforeAll.ts:8]: I kind of like the idea, but I'm changing beforeEach in my PR, to short circuit when the story is aborted, so would be kind of less straight-forward to make this generic:

```ts
  const applyBeforeEach = async (context: StoryContext<TRenderer>): Promise<CleanupCallback[]> => {
    const cleanupCallbacks = new Array<() => unknown>();
    for (const beforeEach of [
      ...normalizeArrays(projectAnnotations.beforeEach),
      ...normalizeArrays(componentAnnotations.beforeEach),
      ...normalizeArrays(storyAnnotations.beforeEach),
    ]) {
      if (context.abortSignal.aborted) return cleanupCallbacks;
      const cleanup = await beforeEach(context);
      if (cleanup) cleanupCallbacks.push(cleanup);
    }
    return cleanupCallbacks;
  };
  ```
I would say, leave this to beforeAll for now.

File: code/lib/types/src/modules/story.ts (hunk 6 above)
Comment (ghengeveld) [reply code/lib/types/src/modules/story.ts:50]: 👉 https://github.com/ComponentDriven/csf/pull/96

File: code/renderers/svelte/src/portable-stories.ts (hunk 7 above)
Comment (yannbf) [reply code/renderers/svelte/src/portable-stories.ts:13]: It's a fix that was done in React but not in other renderers, by mistake. To allow users to use preview.ts with both named export or default export formats:

```ts
export const decorators = []
```

```ts
const preview = {
  decorators: []
]
export default preview
```

**Comment by ghengeveld** [comment]
e2e-dev is failing but not because of the work in this PR...

**Review by kasperpeulen — APPROVED** [review]
LGTM!
